hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java

```diff
@@ -54,11 +56,15 @@ private static void moveToNextHour(Calendar calendar) {
 
   public static int getCurrentHour() {
     Tick tick = CurrentHourProvider.tick;
-    if(System.currentTimeMillis() < tick.expirationTimeInMillis) {
+    if (EnvironmentEdgeManager.currentTime() < tick.expirationTimeInMillis) {
       return tick.currentHour;
     }
 
     CurrentHourProvider.tick = tick = nextTick();
     return tick.currentHour;
   }
+
+  protected static void forceUpdateTickForTest() {
```

Comment:
I think this can be done by implementing a special EnvironmentEdge? Where we will return a pre defined sequence of timestamp.

Reply:
Do you mean like this:

>     if (EnvironmentEdgeManager.currentTime() <= 23) {
>       return (int)EnvironmentEdgeManager.currentTime();
>     }

Reply:
Then we can inject a special value.

>     EnvironmentEdgeManager.injectEdge(()->0L);
>     assertEquals(0, CurrentHourProvider.getCurrentHour());
>     EnvironmentEdgeManager.injectEdge(()->15L);
>     assertEquals(15, CurrentHourProvider.getCurrentHour());

Reply:
```
class MockEnvironmentEdge implements EnvironmentEdge {
  private final Queue<Long> times = new ArrayDeque<>();

  public void inject(long timestamp) {
    times.addLast(timestamp);
  }

  public long currentTime() {
    return times.pollFirst();
  }
}

EnvironmentEdgeManager.injectEdge(new MockEnvironmentEdge());
```

Something like this.

Reply:
A bit confusion of why we need a queue?

Reply:
You can use any Collections you want. The point here, is that you can inject multiple values at once, and also you can implement more complicated logic in the currentTime method, so you do not need to add a forceUpdateTickForTest method.

Reply:
Oh, but i think the test failure above is caused by injectEdge in other test, if EnvironmentEdgeManager.currentTime() < tick.expirationTimeInMillis, then it will return the previous currentHour, it is why i want force update the tick.

Reply:
How about the way i mentioned above, a big magic, but seems bring smallest impact.

> if (EnvironmentEdgeManager.currentTime() <= 23) {
>   return (int)EnvironmentEdgeManager.currentTime();
> }

Reply:
Then just make the tick field as package private and add a VisibleForTesting annotation? Just change it directly in UT.

Reply:
Also the nextTick(), let see the qa result for the recent build first.